### DISCUSSION TITLE
Chunk uploading with API

### DISCUSSION THREAD

**Discussion by vectorkovacspeter** [Forceu/Gokapi#135] [Q&A]
Hello,

I use the following js code to make chunk upploading of zip file
My problem is, that the file and links created well, but when I download the file, I got invalid zip file.
The size seems to be right.

Can someone please suggest me a solution, or is this a known bug in the server?

The relevant code:
```javascript
const UploadFileChunked = async (
  fileName,
  filePath,
  fInfo,
  fileSizeInMegaBytes,
  filePassword,
  fileDownloadLimit,
  fileDaysLimit,
) => {
  const fileUUID = uuidv4();
  const chunkSize = 10 * 1024 * 1024; // 5MB
  const totalChunks = Math.ceil(fileSizeInMegaBytes / fileChunkLimitMb);

  console.log({
    chunkSize,
    totalChunks,
    fileSizeInMegaBytes,
    fileChunkLimitMb,
  });

  for (let i = 0; i < totalChunks; i++) {
    const start = i * chunkSize;
    const end = Math.min(start + chunkSize, fInfo.size);
    const chunk = fs.createReadStream(filePath, { start, end });

    const options = {
      method: "POST",
      headers: {
        Accept: "application/json",
        apikey: apiKey,
        "Content-Type": "multipart/form-data",
      },
      rejectUnauthorized: false,
      formData: {
        file: {
          value: chunk,
          options: {},
        },
        uuid: fileUUID,
        filesize: fInfo.size,
        offset: i,
      },
    };

    console.info(`chunk - ${i}`);
    try {
      await requestAsync(`${uri}/api/chunk/add`, options);
    } catch (err) {
      console.error(err);
      Logger.info(`chunk - ${i}`);
      Logger.exception(err);
      throw err;
    }
  }
  const optionsFinal = {
    method: "POST",
    headers: {
      Accept: "application/json",
      apikey: apiKey,
      "Content-Type": "application/x-www-form-urlencoded",
    },
    rejectUnauthorized: false,
    form: {
      uuid: fileUUID,
      filename: fileName,
      filesize: fInfo.size,
      allowedDownloads: fileDownloadLimit,
      expiryDays: fileDaysLimit,
      password: filePassword,
    },
  };

  console.info(`last chunk`);
  console.log(optionsFinal);
  try {
    return await requestAsync(`${uri}/api/chunk/complete`, optionsFinal);
  } catch (err) {
    console.error(err);
    Logger.info("last chunk");
    Logger.exception(err);
    throw err;
  }
};
```

**Comment by Forceu** [comment]
This appears to be a bug on your side. ChatGPT said that you might need to replace `Math.ceil(fileSizeInMegaBytes / fileChunkLimitMb);` with `Math.ceil(fileSizeInMegaBytes / chunkSize);`, which would make sense

**Reply by vectorkovacspeter** [reply]
Thanks for your time to check this, the suggestion really make sense, but unfortunately the end result still the same, the upload finished, the file and the link create on the Gokapi side, its size seems ok, but damaged when download.

The same file works well, when I use the Admin UI to uppload it.

Do you know other opensource JS app or client library for this project? Maybe I can check someones code :)

**Reply by Forceu** [reply]
Have you opened the file with a hexeditor? The part where it the file is corrupted probably has only zeros, as that chunk is missing. Otherwise try this file, set the chunk size to 2 bytes and then have a look which one is missing. 

As far as I know, there are no public JS clients.

[hextest.txt](https://github.com/Forceu/Gokapi/files/14193648/hextest.txt)
